core/src/sigverify_shreds.rs

```diff
@@ -0,0 +1,395 @@
+        thread_pool.borrow().install(|| {
+            batches
+                .into_par_iter()
+                .map(|p| {
+                    p.packets
+                        .par_iter()
+                        .map(|p| verify_shred_cpu(p, slot_leaders).unwrap_or(0))
+                        .collect()
+                })
+                .collect()
+        })
+    });
+    inc_new_counter_debug!("ed25519_shred_verify_cpu", count);
+    rv
+}
+
+fn shred_gpu_pubkeys(
+    batches: &[Packets],
+    slot_leaders: &HashMap<u64, Pubkey>,
+    recycler_offsets: &Recycler<TxOffset>,
+    recycler_pubkeys: &Recycler<PinnedVec<Pubkey>>,
+) -> (PinnedVec<Pubkey>, TxOffset) {
+    //TODO: mark Pubkey::default shreds as failed after the GPU returns
+    assert_eq!(slot_leaders.get(&std::u64::MAX), Some(&Pubkey::default()));
+    let slots: Vec<Vec<u64>> = PAR_THREAD_POOL.with(|thread_pool| {
+        thread_pool.borrow().install(|| {
+            batches
+                .into_par_iter()
+                .map(|p| {
+                    p.packets
+                        .iter()
+                        .map(|packet| {
+                            let slot_start = size_of::<Signature>() + size_of::<ShredType>();
+                            let slot_end = slot_start + size_of::<u64>();
+                            let slot: Option<u64> =
+                                deserialize(&packet.data[slot_start..slot_end]).ok();
+                            if slot.is_some() && slot_leaders.get(&slot.unwrap()).is_some() {
+                                slot.unwrap()
+                            } else {
+                                std::u64::MAX
+                            }
+                        })
+                        .collect()
+                })
+                .collect()
+        })
+    });
+    let mut keys: HashMap<Pubkey, Vec<u64>> = HashMap::new();
+    for batch in slots.iter() {
+        for slot in batch.iter() {
+            let key = slot_leaders.get(slot).unwrap();
+            keys.entry(*key).or_insert(vec![]).push(*slot);
+        }
+    }
+    let mut pubkeys: PinnedVec<Pubkey> = recycler_pubkeys.allocate("shred_gpu_pubkeys");
+    let mut slot_to_key_ix = HashMap::new();
+    for (i, (k, slots)) in keys.iter().enumerate() {
+        pubkeys.push(*k);
+        for s in slots {
+            slot_to_key_ix.insert(s, i);
+        }
+    }
+    let mut offsets = recycler_offsets.allocate("shred_offsets");
+    slots.iter().for_each(|packet_slots| {
+        packet_slots.iter().for_each(|slot| {
+            offsets.push((slot_to_key_ix.get(slot).unwrap() * size_of::<Pubkey>()) as u32);
+        });
+    });
+    //HACK: Pubkeys vector is passed along as a `Packets` buffer to the GPU
+    //TODO: GPU needs a more opaque interface, which can handle variable sized structures for data
+    //Pad the Pubkeys buffer such that it is bigger than a buffer of Packet sized elems
+    let num_in_packets =
```

Comment:
I think this calculation is nicely-written! I found no problem after I thought hard how this calculation can resist to an attack where malicious sender sends packets with bad slot resulting in a corner case of `pubkeys` just being `[std::u64::MAX]`!

Reply:
pubkeys is generated locally, it's not an external value.

Reply:
@sakridge Thanks for additional explanation. My original concern is that this padding calculation might go wrong when `pubkeys` with having **only a single element**. On the other hand, `pubkeys` will usually be a unique set of pubkeys of slot leaders usually greater than 1 elements.

As you said, the actual value of `pubkeys`' elements are not external value indeed, however, I think _the length of `pubkeys`_ are externally controllable via crafted series of packets with bad slots. So I pondered on it a bit. And I left a remark saying that no problem is found. :)